[CI] RelocationIT.testIndexAndRelocateConcurrently fails on master intake

https://gradle-enterprise.elastic.co/s/k63bx3zx5aeg4

REPRODUCE WITH: ./gradlew ':server:integTest' --tests "org.elasticsearch.recovery.RelocationIT.testIndexAndRelocateConcurrently" -Dtests.seed=B205D9E1D369E428 -Dtests.security.manager=true -Dtests.locale=sr-BA -Dtests.timezone=America/Goose_Bay -Dcompiler.java=12 -Druntime.java=11

The failure doesn't reproduce for me

Relevant extract from the console log: 
```
java.lang.AssertionError: Unexpected ShardFailures: [[test][3] failed, reason [[test/jPGCBDDYR36D8UlPCUoAYQ][[test][3]] org.elasticsearch.action.support.broadcast.BroadcastShardOperationFailedException:
at org.elasticsearch.action.support.replication.TransportBroadcastReplicationAction.finishAndNotifyListener(TransportBroadcastReplicationAction.java:166)
...

Caused by: org.elasticsearch.transport.RemoteTransportException: [node_t3][127.0.0.1:33581][indices:admin/refresh[s][r]]
--
Caused by: java.lang.IllegalStateException: store is already closed can't increment refCount current count [0]
...

Expected: <0>
--
but: was <1>
at __randomizedtesting.SeedInfo.seed([B205D9E1D369E428:6D34D86CCF6CFB8B]:0)
at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertNoFailures(ElasticsearchAssertions.java:326)
at org.elasticsearch.test.ESIntegTestCase.indexRandom(ESIntegTestCase.java:1367)
at org.elasticsearch.test.ESIntegTestCase.indexRandom(ESIntegTestCase.java:1272)
at org.elasticsearch.test.ESIntegTestCase.indexRandom(ESIntegTestCase.java:1255)
at org.elasticsearch.test.ESIntegTestCase.indexRandom(ESIntegTestCase.java:1231)
at org.elasticsearch.recovery.RelocationIT.testIndexAndRelocateConcurrently(RelocationIT.java:492)
```

--- a/server/src/main/java/org/elasticsearch/index/engine/Engine.java
+++ b/server/src/main/java/org/elasticsearch/index/engine/Engine.java
@@ -1156,8 +1156,7 @@ public abstract class Engine implements Closeable {
         if (failure != null) {
             maybeDie(reason, failure);
         }
-        if (failEngineLock.tryLock()) {
-            store.incRef();
+        if (failEngineLock.tryLock() && store.tryIncRef()) {
             try {
                 if (failedEngine.get() != null) {
                     logger.warn(() ->

--- a/server/src/main/java/org/elasticsearch/index/engine/Engine.java
+++ b/server/src/main/java/org/elasticsearch/index/engine/Engine.java
@@ -1156,7 +1156,7 @@ public abstract class Engine implements Closeable {
         if (failure != null) {
             maybeDie(reason, failure);
         }
-        if (failEngineLock.tryLock() && store.tryIncRef()) {
+        if (failEngineLock.tryLock()) {
             try {
                 if (failedEngine.get() != null) {
                     logger.warn(() ->
@@ -1178,11 +1178,19 @@ public abstract class Engine implements Closeable {
                     // on the same node that we don't see the corrupted marker file when
                     // the shard is initializing
                     if (Lucene.isCorruptionException(failure)) {
-                        try {
-                            store.markStoreCorrupted(new IOException("failed engine (reason: [" + reason + "])",
-                                ExceptionsHelper.unwrapCorruption(failure)));
-                        } catch (IOException e) {
-                            logger.warn("Couldn't mark store corrupted", e);
+                        if (store.tryIncRef()) {
+                            try {
+                                store.markStoreCorrupted(new IOException("failed engine (reason: [" + reason + "])",
+                                    ExceptionsHelper.unwrapCorruption(failure)));
+                            } catch (IOException e) {
+                                logger.warn("Couldn't mark store corrupted", e);
+                            } finally {
+                                store.decRef();
+                            }
+                        } else {
+                            logger.warn(() ->
+                                    new ParameterizedMessage("tried to mark store as corrupted but store is already closed. [{}]", reason),
+                                failure);
                         }
                     }
                     eventListener.onFailedEngine(reason, failure);
@@ -1191,8 +1199,6 @@ public abstract class Engine implements Closeable {
                 if (failure != null) inner.addSuppressed(failure);
                 // don't bubble up these exceptions up
                 logger.warn("failEngine threw exception", inner);
-            } finally {
-                store.decRef();
             }
         } else {
             logger.debug(() -> new ParameterizedMessage("tried to fail engine but could not acquire lock - engine should " +

--- a/server/src/test/java/org/elasticsearch/recovery/RelocationIT.java
+++ b/server/src/test/java/org/elasticsearch/recovery/RelocationIT.java
@@ -490,7 +490,6 @@ public class RelocationIT extends ESIntegTestCase {
             docs[i] = client().prepareIndex("test").setId(id).setSource("field1", English.intToEnglish(numDocs + i));
         }
         indexRandom(true, docs);
-        numDocs *= 2;
 
         logger.info(" --> waiting for relocation to complete");
         ensureGreen(TimeValue.timeValueSeconds(60), "test"); // move all shards to the new nodes (it waits on relocation)

